text/0000-exit-codes.md

```diff
@@ -0,0 +1,117 @@
+
+    /// A network error occurred.
+    NetworkError = 5,
+
+    /// The requested executable could not be run.
+    ExecutionFailure = 126,
+
+    /// The requested executable is not available.
+    ExecutableNotFound = 127,
+}
+```
+
+### Grouping
+
+Errors will be grouped according to basic cause, with thought given to how external tooling might choose to react to the problem.  Doing so saves external tools from trying to parse Notion error messages in order to make these decisions.  For example, an external tool receiving an exit code representing an error parsing the command line should absolutely not retry the operation (and may have a bug).  An exit code indicating that no matching version is available indicates a potentially non-permanent error, but one which shouldn't be retried without first using another command to change Notion's current state.  An exit code indicating a network error signals a kind of transient problem which might be desirable to retry after a short delay rather than immediately giving up entirely.
+
+### Documentation
+
+Customer-facing documentation will also need to be generated, either by hand or by using a purpose-made tool.  (Rustdoc alone is insufficient for this purpose, as it does not expose the enum's integer values.)  The documentation for the above example might look as follows.  (Note that the descriptions provided to the user is the same as the ones provided to developers.)
+
+| Code | Description
+| ---: | :--
+|    0 | No error occurred.
+|    1 | An unknown error occurred.
+|    3 | An invalid combination of command-line arguments was supplied.
+|    4 | No match could be found for the requested version string.
+|    5 | A network error occurred.
+|  126 | The requested executable could not be run.
+|  127 | The requested executable is not available.
+
+# Critique
+[critique]: #critique
+
+The primary downside introduced by this proposal is the loss of error granularity.  By grouping multiple errors into the same exit code, external tooling loses the ability to differentiate between those errors without capturing and parsing Notion's output.  However, it doesn't seem realistic to provide a separate exit code for each error due to the limited nature of the former and the limitless potential of the latter.  🙂
+
+Another issue is the communication of exit codes.  We want to publish the list as part of Notion's standard, user-facing documentation, but this requires either manually keeping the documentation in sync with the source or writing a tool to do so for us.  The alternative is to point interested parties directly to the Notion source, which isn't ideal.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Is a process needed for 'vetting' new exit codes, to prevent duplication?  If so, is the RFC process appropriate?  Alternatively, is code review sufficient?
+- Should we engage npm/Yarn to improve error categorization (either through parsable output or diversifying exit codes)?
```

Comment:
Yarn does have the ability to generate JSON formatted output, although currently you have to choose one or the other. I have reached out to some folks involved with Yarn to see if it'd be possible to "tee" the output via a config file or environment variable, so that a developer could choose whatever output they want at stdout/stderr, but we could specify a side file that also receives the output in JSON format.